### ISSUE TITLE
About the use of OpenSSL library functions

### ISSUE THREAD

**Issue by johnflj** [eclipse-cyclonedds/cyclonedds#1535]
@eboasson  HI, eGod:
I encountered the following problem:
 In ubuntu22, the default openssl version is updated to 3.0, and version 3.0 deprecates SSL_get_peer_certificate.
because dds_report_tls_version calls SSL_get_peer_certificate, 
it will cause The compiler reports the following error: /usr/bin/ld: ... /... /... /lib/libddsc.so.0.11.0: undefined reference to `SSL_get_peer_certificate'.
The OpenSSL web page describes this as follows.
SSL_get0_peer_certificate() and SSL_get1_peer_certificate() were added in 3.0.0. 
SSL_get_peer_certificate() was deprecated in 3.0.0.

How to handle the dds_report_tls_version function to achieve compatibility between different versions of OpenSSL and to avoid compilation errors due to good versions of openssl?
The openssl reference link is as follows.
https://www.openssl.org/docs/man3.1/man3/SSL_get_peer_certificate.html

**Comment by eboasson** [comment]
Hi @johnflj, that's a nasty little detail in OpenSSL. I haven't ran into it myself, but I think what you suggest is the most sensible thing: just call `SSL_get1_peer_certificate` if using OpenSSL 3.0.

I have had to do some OpenSSL version-specific code before, like in:
https://github.com/eclipse-cyclonedds/cyclonedds/blob/38aeacd81ba4f6818ff53d5226b7075b343a60c2/src/core/ddsi/src/ddsi_ssl.c#L134

it seems to me that we can do actually the same here, just with a different OpenSSL version number. A bit like:
```C
  char issuer[256], subject[256];
#if OPENSSL_VERSION_NUMBER < 0x30000000L
  peercert = SSL_get_peer_certificate (ssl)
#else
  peercert = SSL_get_peer1_certificate (ssl)
#endif
 X509_NAME_oneline (X509_get_issuer_name (peercert), issuer, sizeof (issuer));
  X509_NAME_oneline (X509_get_subject_name (peercert), subject, sizeof (subject));
  GVTRACE ("tcp/ssl %s %s issued by %s [%s]\n", oper, subject, issuer, SSL_get_version (ssl));
  SSL_free (peercert);
```

The `SSL_free` is to fix a memory leak that currently exists, from reading the man pages. If it'd be OpenSSL 3-only, one would ideally use `SSL_get0_peer_certificate`, but 1.1.1 is still a thing ...

**Comment by johnflj** [comment]
Thanks, eGod.

**Comment by eboasson** [comment]
Quick question @johnflj: does your closing this as completed mean that my snippet did the trick without modification? ('t Was completely untested, so it is not supposed to work! 🤓) Because then I should make a PR.

**Comment by johnflj** [comment]
@eboasson Hi,eGod
This is the model that I modified locally.It works, but the version number of version 3.0.0 and above is not defined by a macro in the header file, so there was a problem debugging at first。

**Comment by eboasson** [comment]
Thanks @johnflj!

I've just created an Ubuntu 22 VM because of some ROS-related things, so with a bit of luck I can reproduce your original problem there. It needs fixing, it is not good if you have to downgrade to Ubuntu 20 or to OpenSSL 1.1.x ...

**Comment by johnflj** [comment]
@eboasson 
Hi，eGod，Currently, is there an easy way to support openssl version 3.0.2?

> Thanks @johnflj!
> 
> I've just created an Ubuntu 22 VM because of some ROS-related things, so with a bit of luck I can reproduce your original problem there. It needs fixing, it is not good if you have to downgrade to Ubuntu 20 or to OpenSSL 1.1.x ...

**Comment by eboasson** [comment]
@johnflj I should've come back to you earlier: I did try it, but I could not reproduce it. That was with a clean Ubuntu 22.10 server install in an aarch64 VM, and building after doing `apt install build-essential cmake libssl-dev`. Is it possible that your environment has a mix of versions and that that is causing confusion in the build?

**Comment by johnflj** [comment]
> @johnflj我应该早点回到你身边：我确实尝试过，但我无法重现它。那是在 aarch22 VM 中安装干净的 Ubuntu 10.64 服务器，并在执行 .您的环境是否有可能混合了版本，这会导致构建中的混乱？`apt install build-essential cmake libssl-dev`

@eboasson
Hi，eGod，
The version of ubuntu I am using is Ubuntu 22.04.1 LTS (GNU/Linux 5.15.0-58-generic x86_64). The version of openssl that comes with the system is OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)

**Comment by johnflj** [comment]
> I had 3.0.5 on that Ubuntu 22.10 one, but I had to redo it on Ubuntu 22.04 because building ROS 2 on 22.10 is a no-go at the moment (it may have been fixed since), and so now I have OpenSSL 3.0.2. It builds and runs all tests without problems.
> 
> So I am thinking there's got to be a mix-up of versions. Just about the only thing I can think of is that you someone (also) have OpenSSL header files for, e.g., 1.1.x, and that the Cyclone build is picking that wrong set of header files.

@eboasson 
Hi, eGod,
The scenario you describe is the most likely one. OpenSSL 1.1.1X libraries are not included by default on 22.04 ubuntu systems, unless you install and import them. You can control the query priority of dynamic libraries by specifying the path to the dynamic libraries (LD_LIBRARY_PATH) at compile time. You can also use the ldd executable to verify that the dynamic library file linked by the executable is the system specified OpenSSL version 3.0.X.

Translated with www.DeepL.com/Translator (free version)
